### DISCUSSION TITLE
Issue using projen with private npm registry such as CodeArtifact.

### DISCUSSION THREAD

**Discussion by mtahir-cloudkitect** [projen/projen#3901] [Q&A]
I have an AwsCdkTypeScriptApp project that needs to pull some dependencies from a private code artifact repository. 
I tried the following configuration

`project.npmrc.addRegistry('https://code_artifact_repo_url/npm/repository/')
project.npmrc.addConfig('//code_artifact_repo_url/npm/repository:_authToken', '$(aws codeartifact get-authorization-token --domain domain_name --domain-owner account_id --query authorizationToken --output text --region us-east-1)')
`

My intent was that it would execute command $( ) and use its value for populating authToken. But it outputs the string as is.

Basically I need to handle 2 scenarios, 
1. Developers working on local machines, downloading the dependencies
2. Build pipelines, downloading the dependencies

Has anyone found a solution of using projen with private npm registry such as code artifact?

**Comment by mrgrain** [comment]
Projen has some support for CodeArtifact build in: https://projen.io/docs/publishing/publisher-component#publishing-to-aws-codeartifact

Did you try that yet?

**Comment by mrgrain** [comment]
Does `$( )` work without projen? My starting point is always, how would I do this without projen and then once that is working you can look at using projen for it.

Personally I don't think NPM `_authToken`s should go into a project's `.npmrc` file at all. They should live in the user config `~/.npmrc`. You'd need to document for your devs how to do that, and you might provide them with a `projen login` command. But ultimately the individual dev would be responsible to log in. In CI you'll of course need to configure this with commands and/or env variables.

**Comment by mtahir-cloudkitect** [comment]
Thank you for the response, projen's built in support for CodeArtifact is for publishing artifacts, not for downloading artifacts. I sort of made it work by adding the following line to projenrc.ts file 
project.npmrc.addConfig('@org:registry', 'https://codeartifact-url/artifacts/');

And then as you mentioned, asking the developers to login to code artifact so that the token gets added to their ~/.npmrc file. 

On the CI side I added support for it in the deployable cdk project 
https://github.com/cloudkitect/deployable-cdk-application

Now the trouble is in generating python distribution. Python engine needs to know how to download dependencies from codeartifact. So far I have tried 
export TWINE_REPOSITORY_URL = https://codeartifact-url/artifacts/
export TWINE_USERNAME=aws
export TWINE_PASSWORD=`aws codeartifact get-authorization-token --domain DOMAIN --domain-owner ACCOUNT --query authorizationToken --output text --region us-east-1`

Still have not found the solution. :)

**Reply by mrgrain** [reply]
Yeah I'm not super familiar with Python. This might help https://docs.aws.amazon.com/codeartifact/latest/ug/python-configure-pip.html
